File: src/disjoint-sets-union-find/utils/disjoint_sets.ts
export default class DisjointSets {
  private parent: number[];
  private size: number[];
  constructor(n: number) {
    this.parent = Array.from({ length: n + 1 }, (_, i) => i);
    this.size = Array(n + 1).fill(0);
  }

  public findRoot(a: number): number {
    if (this.parent[a] === a) {
      return a;
    }
    this.parent[a] = this.findRoot(this.parent[a]);
    return this.parent[a];
  }

  public unionSets(a: number, b: number): void {
    const rootA = this.findRoot(a);
    const rootB = this.findRoot(b);
    if (rootA !== rootB) {
      if (this.size[rootA] < this.size[rootB]) {
        this.parent[rootA] = rootB;
        this.size[rootB] += this.size[rootA];
      } else {
        this.parent[rootB] = rootA;
        this.size[rootA] += this.size[rootB];
      }
    }
  }

  public isConnected(a: number, b: number): boolean {
    return this.findRoot(a) === this.findRoot(b);
  }
}

File: src/disjoint-sets-union-find/utils/disjoint_sets.py
class DisjointSets:
    def __init__(self, n: int):
        self.parent = [i for i in range(n + 1)]
        self.size = [0] * (n + 1)

    def find_root(self, a: int) -> int:
        if self.parent[a] == a:
            return a
        self.parent[a] = self.find_root(self.parent[a])
        return self.parent[a]

    def union_sets(self, a: int, b: int):
        rootA = self.find_root(a)
        rootB = self.find_root(b)
        if rootA != rootB:
            if self.size[rootA] < self.size[rootB]:
                self.parent[rootA] = rootB
                self.size[rootB] += self.size[rootA]
            else:
                self.parent[rootB] = rootA
                self.size[rootA] += self.size[rootB]

    def is_connected(self, a: int, b: int) -> bool:
        return self.find_root(a) == self.find_root(b)
